---
title: "RaMWAS parameters"
author: "Andrey A Shabalin"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true # table of content true
vignette: >
  %\VignetteIndexEntry{RaMWAS parameters}
  %\VignetteEngine{knitr::rmarkdown}
---

# Initialize RaMWAS Parameters

There are several ways to initialize the parameters for calling any RaMWAS pipeline function.
The parameters can be stored in an R list like this:

```{r eval=FALSE}
param = list(
	dirbam = "/ramwas_project/bams/",
	dirproject = "/ramwas_project/",
	filebamlist = "/ramwas_project/000_list_of_files.txt",
	scoretag = "AS",
	minscore = 100,
	cputhreads = 8,
	filecpgset = "/ramwas/cpgsets/cpgset_hg19_SNPS_at_MAF_0.05.rds",
	maxrepeats = 3,
	maxfragmentsize=200,
	minfragmentsize=50,
	bamnames = NULL,
	filebam2sample = "/ramwas_project/bam2sample.txt",
	minavgcpgcoverage = 0.3,
	minnonzerosamples = 0.3,
	covfile = "cov.txt",
	chrkeep = 1:22
);
```

Alternatively, the parameters can be set in a separate R code file, which is processed into a list as above by `parametersFromFile` function.
The R code file can contain lines like this:

```{r eval=FALSE}
### R parameter file
dirbam = "/ramwas_project/bams/"
dirproject = "/ramwas_project/"
filebamlist = "/ramwas_project/000_list_of_files.txt"
scoretag = "AS"
minscore = 100

### platform dependent part
if(.Platform$OS.type == "windows") {
	filecpgset='C:/RaMWAS/CpG_set/cpgset_hg19_SNPS_at_MAF_0.05.rds'
} else {
	filecpgset='/computing_cluster/ramwas/cpgset_hg19_SNPS_at_MAF_0.05.rds'
}


```

# Explanation of all parameters

## Parameters pointing to files

### File with BAM names

Parameter 'filebamlist', if defined, must point to a text file with one BAM file name per line. BAM file names can contain path, relative or absolute.
Such file may looks like this.
```{r eval=FALSE}
batch1/b1sample1.bam
batch1/b1sample2.bam
batch2/b2sample1.bam
batch2/b2sample2.bam
batch2/b2sample3.bam
batch4/sample4.bam
```

This file is then loaded into `bamnames` parameter.
Note that BAM names must be all different. Namely, such list of BAMS is **NOT** allowed:
```{r eval=FALSE}
batch1/sample1.bam
batch1/sample2.bam
batch2/sample1.bam
batch2/sample2.bam
```


### File with CpG locations

RaMWAS calculates coverage at a prepared set of CpG locations. The set depends on the reference genome and usually includes CpGs created by common SNPs.
The set of CpGs is stored as a `list` with one sorted vector of CpG locations per chromosome.

```{r CpGsetExample}
cpgset = list( chr1 = c(12L, 57L, 123L),
               chr2 = c(45L, 95L, 99L, 111L),
               chr3 = c(22L, 40L, 199L, 211L) );
```

Parameter `filecpgset` must be a file storing a set of CpG, in the format described above, saved using `saveRDS` function.
Similarly, parameter `filenoncpgset`, if defined, must point to a file storing 'good' locations away from any CpGs. Here, good means not heaving major alignment issues (like a middle of centromere.

### File with bam 2 sample matching information

Parameter `filebam2sample` must point to a file with information which samples correspont to which BAMs.

The file here defines samples `sample1`, `sample2`, and `sample3`. Sample1 is assigned reads from bams `b1sample1` and `b2sample1`.
```{r eval=FALSE}
sample1=b1sample1,b2sample1
sample2=b1sample2,b2sample2
sample3=b2sample3
sample4
```
If a sample has only one BAM and the BAM has the same name as the sample, the corresponding line can contain just that name (see `sample4` above).

The `filebam2sample` file is scanned into `bam2sample` parameter.

### File with covariates

The parameter `filecovariates`, if defined, must point to a text file which can be read with `read.table`.
If the file has extension ".csv", it is assumed to be comma separated, otherwise - tab separated.

The file `filecovariates` is read into `covariates` parameter.

## directory parameters











