---
title: "RaMWAS pipeline parts"
author: "Andrey A Shabalin"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true # table of content true
vignette: >
  %\VignetteIndexEntry{RaMWAS pipeline parts}
  %\VignetteEngine{knitr::rmarkdown}
---

# Creating artificial data set

The following function generates an artificial data set which would be useful for illustrating the steps of RaMWAS pipeline.

```{r generateData}
{
	suppressPackageStartupMessages(library(ramwas))
	dir = paste0(tempdir(), '/simulated_project');
	ramwas0createArtificialData(dir, nsamples = 20, nreads = 1e6, verbose=FALSE)
	cat("Artificial project files created in:", "\n", dir, "\n")
	message("The generated files and directories are:");
	message(paste(list.files(dir), collate="\n"))
}
```

# Project directory structure

Typically RaMWAS pipeline creates the following directory structure.

* Project directory (`dirproject` parameter, current directory by default)
    * Rbam files (`dirrbam` parameter or 'rds_rbam' by default)
    * R QC files (`dirrqc` paramter or 'rds_qc' by default)
    * QC plots and figures ('dirqc' parameter or 'qc' by default)
    * Normalized coverage, 1 CpG per column (`dircoveragenorm` paramteter or 'coverage_norm_123' by default, where 123 is the number of samples)
        * MWAS subdirectory, named based on the phenotype, number of PCs and covariates used in analysis
            * 10-fold cross-validation and multi-marker prediction rate estimate.

# Pipeline parameters

Each step of the pipeline accepts a list of named values as a parameter. 
We set the initial set of parameters for the artificial data set as follows:
```{r parameters}
param = list(
    dirproject = dir,
    dirbam = paste0(dir,"/bams"),
    scoretag = "mapq",
    minscore = 4,
    filebamlist = "bam_list.txt",
    filebam2sample =  "bam_list.txt",
    filecovariates = "covariates.txt",
    modeloutcome = "casecontrol",
    filecpgset = paste0(dir,"/Single_chromosome.rds"),
    minfragmentsize = 75,
    maxfragmentsize = 250,
    cputhreads = 2,
    minavgcpgcoverage = 0.3,
    minnonzerosamples = 0.3
)
```

See `r vignette("RaMWAS_param")` for full information on the available parameters.
            
# Pipeline, 6 major parts

## Scan BAM files, filtering reads, calculation of QC measures, saving the read start positions.

The BAM files must be located in `dirbam` and listed in `filebamlist`. The reads are filtered out if they have tag `scoretag` value less than `minscore`. The the read start information and other QC measures is then recorded. The scanning is done in parallel using at most `cputhreads` threads.

```{r scan bams}
{
	suppressMessages({
	ramwas1scanBams(param)
	})
	message("The generated files and directories are:");
	message(paste(list.files(dir), collate="\n"))
}
```

**R function.** This task can be run with a single call of the `ramwas1scanBams` or separate runs of `pipelineProcessBam` for each BAM file.
The `ramwas1scanBams` function starts a local cluster To process BAMs in parallel.
The `pipelineProcessBam` can be used if the user wants to manually spread the task across the nodes of a computing cluster.

**Parameters**

* The alignments in the BAM files are filtered by alignment score (often MAPQ or AS tag, set by `scoretag`), the reads with the score tag below `minscore` are discarded.
* BAM file names are taken from either `filebamlist` file or `bamnames` paramter.
* BAMs are searcher for in `dirbam` directory (or relative to it).
* Reads are filtered by `scoretag` tag in the BAM file (typically "MAPQ" or "AS"). Reads with the score tag below `minscore` are excluded.
* Up to `cputhreads` parallel R jobs are used to process BAMs (if `cputhreads` > 1).
* The QC (Quality Control) measures are calculated using a number of parameters, including `filecpgset`, `filenoncpgset`, `maxfragmentsize`.

**Note**

* BAMs can be located in different directories, but must have unique names.
* If a BAM has already been scanned, it is not processed again.
* The information extracted from the BAMs is stored in R format in "rds_rbam" subdirectory of the Project directory. 
* The QC information in R format is stored in "rds_qc" directory. 
* The QC plots and figures are stored in "qc" directory.
* See `vignette("RaMWAS_param")` for information about the QC measures.

## Collect QC measures from individual BAMs, summarizing by samples, recording summary and plots.

Next we collect QC information across BAMs, collect it by samples (if multiple BAMs correspond to one sample), and estimate fragment size distribution from the read start locations around isolated CpGs.
```{r collectQC}
{
	suppressMessages({
	ramwas2collectqc(param)
	})
	message("The generated files and directories in QC directory are:");
	message(paste(list.files(paste0(dir,"/qc")), collate="\n"))
}
```

**Parameters**

* Multiple BAMs can contain information from a single sample. The bam to sample matching information is taken from `filebam2sample` file or directly from `bam2sample` character vector.

**Note**

* All QC measures are summarized by BAM, by sample, and combined in one total metric across all bams.
* The resulting plots and tables are recored in `dirqc` subdirectories "summary_bams", "summary_bams_in_bam2sample", "summary_by_sample", and "summary_total".
* The fragment size distribution is estimated using all BAMs in `bam2sample` parameter (if defined) and saved in "Fragment_size_distribution.txt" file in the project directory.


## Calculate coverage matrix for all CpGs across all samples, filter CpGs by coverage, normalize samples to the same sample average.

Next, we create a coverage matrix, normalized to have the same average for each sample.
The coverage matrix is stored in big file as a `filematrix` object. See `filematrix` package for format details.
Again, the task is parallelized across `cputhreads` threads.
```{r normCoverage}
{
	suppressMessages({
	ramwas3NormalizedCoverage(param)
	})
	message("The generated files and directories in coverage directory are:");
	message(paste(list.files(paste0(dir,"/coverage_norm_20")), collate="\n"))
}
```

**Parameters**

* The coverage matrix is calculated for the samples listed in `bam2sample`. 
* If `filecovariates` or `covariates` is defined, only the samples in the covariate file are used.
* The coverage matrix is created in `dircoveragenorm` directory. It is "coverage_norm_1234" by default, where 1234 is the number of samples.
* The coverage is calculated for CpGs in `filecpgset`.
* CpGs with less than `minavgcpgcoverage` average coverage are excluded (dafault is 0.3).
* CpGs with fewer than `minnonzerosamples` samples with nonzero coverage are excluded (default is 0.3, meaning a CpG is preserved if at least 30\% of samples must have non-zero coverage).
* Up to `cputhreads` parallel R jobs are used for calculation of raw coverage (This is a CPU intensive task).
* Up to `diskthreads` parallel R jobs are used for CpG filtering and sample normalization  (This is a disk heavy task).
* The temporary files are located in `dirtemp` directory, which is "temp" subdirectory of `dircoveragenorm` by default.
* For better performance for large projects, set `dirtemp` to a location on another hard drive.

**Note**

* The coverage matrix is stored as a `filematrix` named "Coverage" (see `filematrix` package).
* Filematrices can be accessed (by parts) as usual R matrices without the need to load the whole filematrix in memory.
* The locations of CpG that passed the filter are saved in filematrix "CpG_locations", which has two columns indicating chromosome number and position. The chomosome are numbered as in `filecpgset` (chromosome names are also saved in "CpG_chromosome_names.txt".

## Principal Component Analysis

We run Principal Component Analysis with two goals in mind.
First, we aim to capture main directions of variation in the data and correct for them in the next steps of the analysis.
Second, the sample loadings of the first PCs can indicate highly irregular samples which should be excluded.

```{r pca}
{
	suppressMessages({
	ramwas4PCA(param)
	})
	message("The generated files and directories in PCA directory are:")
	message(paste(list.files(paste0(dir,"/coverage_norm_20/PCA_00_cvrts")), collate="\n"))
	message("Correlation of top PCs with covariates are:");
	head( read.table(paste0(dir,"/coverage_norm_20/PCA_00_cvrts/PC_vs_covariates_corr.txt"), sep = "\t", header = TRUE))
	message("P-values for the associuation of top PCs with covariates are:");
	head( read.table(paste0(dir,"/coverage_norm_20/PCA_00_cvrts/PC_vs_covariates_pvalue.txt"), sep = "\t", header = TRUE))
}
```

**Parameters**

* The principal components are calculated after removing the effect of selected covariates from the data. The covariates can be listed by names in `modelcovariates` parameter. The covariates are taken from `filecovariates` or `covariates`.
* Up to `diskthreads` parallel R jobs are used PCA.
* The PCA output is saved in `dirpca` directory, which by default is named like "PCA_05_cvrts_2d33b43" and located in `dircoveragenorm` directory. The name includes the number of covariates excluded. To differenciate between different sets of, say, 5 covariates it also includes an 8 letter code, which is different for different sets of covariates (most likely).

**Note**

* The sample covariance matrix is saved in "covmat.rds"` file.
* The eigenvalue decomposition of the sample covariance matrix is saved in "eigen.rds" file.
* PC loadings are saved in text format in "PC_loadings.txt".
* PCs are saved in text file "PC_values.txt"
* Up to top 40 PCs are plotted in "PC_plot_covariates_removed.pdf"`. The pdf also contains plots of PC loadings for the first few PCs.
* The correlations of PCs with covariates is recorded in "PC_vs_covariates_corr.txt". 
* The statistical significance of these correlations is recorded in "PC_vs_covariates_pvalue.txt". They can be helpful in selection of covariates for MWAS (Methylome-wide association study). 
		`
		
## Methylome-wide association study (MWAS)

Next we test for association between the outcome variable and methylation coverage.
The testing accounts for the set of covariates and top principal components.
```{r mwas1}
{
	suppressMessages({
	param$modeloutcome = "casecontrol"
	ramwas5MWAS(param)
	})
	message("The generated files and directories in MWAS directories are:");
	message(paste(list.files(paste0(dir,"/coverage_norm_20/PCA_00_cvrts/Testing_casecontrol_0_PCs")), collate="\n"))
	message("The top MWAS findings are:");
	head( read.table(paste0(dir,"/coverage_norm_20/PCA_00_cvrts/Testing_casecontrol_0_PCs/Top_tests.txt"), sep = "\t", header = TRUE))
	qqPlotFast(fm.load(paste0(dir,"/coverage_norm_20/PCA_00_cvrts/Testing_casecontrol_0_PCs/Stats_and_pvalues"))[,3])
	title("Case-control status vs. methylation coverage")
}
```

The testing can easily be repeated for another outcome variable (age):
```{r mwas2}
{
	suppressMessages({
	param$modeloutcome = "age"
	ramwas5MWAS(param)
	})
	message("The generated files and directories in MWAS directories are:");
	message(paste(list.files(paste0(dir,"/coverage_norm_20/PCA_00_cvrts/Testing_age_0_PCs")), collate="\n"))
	message("The top MWAS findings are:");
	head( read.table(paste0(dir,"/coverage_norm_20/PCA_00_cvrts/Testing_age_0_PCs/Top_tests.txt"), sep = "\t", header = TRUE))
	qqPlotFast(fm.load(paste0(dir,"/coverage_norm_20/PCA_00_cvrts/Testing_age_0_PCs/Stats_and_pvalues"))[,3])
	title("Age vs. methylation coverage")
}
```

**Parameters**

* The phenotype of interest is named by `modeloutcome` parameter (taked from `covariates`).
* As in PCA, `modelcovariates` lists covariates to correct for in MWAS.
* The analysis can also include several top PCs as covariates. The number of PCs is set by `modelPCs`.
* The MWAS output is saved in `dirmwas` directory, by default named like "Testing_Age_2_PCs"` where "Age" with the phenotype of interest and 2 is the number of PCs included as covariates.
* Up to `diskthreads` parallel R jobs are used MWAS.
* The title of the qqplot can be set with the parameter `qqplottitle`.

**Note**

* The testing is performed using linear regression model.
* The t-statistics, p-values and q-values for all CpGs are recorded in "Stats_and_pvalues" filematrix. 
* The row of "Stats_and_pvalues" match those of "CpG_locations" filematrix located 2 directories up.
* The QQ-plot of all p-values, with an estimate of inflation parameter lambda is saved in "QQ_plot.pdf".
* Top finding (with p-values below `toppvthreshold`) are annotated with gene names and recorded in "Top_tests.txt".

## Cross Validation with multi-marker elastic net model.

RaMWAS can perform N-fold cross-validation and assess performance of elastic net applied to the top MWAS sites.
```{r cv}
{
	suppressMessages({
	suppressWarnings({
	param$modeloutcomes = "age"
	ramwas6crossValidation(param)
	param$mmncpgs = 10
	ramwas7multiMarker(param)
	param$mmncpgs = 100
	ramwas7multiMarker(param)
	})
	})
	message("The generated files and directories in Multi-marker directory are:");
	message(paste(list.files(paste0(dir,"/coverage_norm_20/PCA_00_cvrts/Testing_age_0_PCs/CV_10_folds")), collate="\n"))
}
```

**R functions.** RaMWAS pipeline can estimate predictive power of elastic net applied to top CpGs from MWAS.
To have an unbiased estimate of predictive power we use cross validation approach.
First, `ramwas6crossValidation` runs MWAS 10 times, each time excluding different 10\% of the data. \cr
Then, `ramwas7multiMarker` trains elastic net on the top CpGs from each of 10 MWASes and predicts the phenotype of interest for the remaing 10\%.
This procedure is then applied 10 times to have a prediction for all samples.

**Parameters**

* By default, 10-fold cross validatidation is performed. The number of folds can be set by `cvnfolds` parameter.
* The split of samples into folds is random. For consistency across runs, the random seed can be set by `randseed`. It is set to 18090212 by default.
* The number of CpGs used to by elastic net is set by `mmncpgs` parameter. It is 1,000 by default.
* The elastic net parameter labmda can be set by `mmalpha` parameter. It is 0 by default.
* The output is saved in `dircv` directory, by default named like "CV_10_folds" and located in `dirmwas` directory.

## Cleanup

Here we remove all the files created by the code chunks above.
```{r clean}
unlink(dir, recursive=TRUE)
```