---
title: "RaMWAS data structures"
author: "Andrey Shabalin"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true # table of content true
    theme: united
vignette: >
  %\VignetteIndexEntry{Filematrix overview}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Loading and Saving RaMWAS objects

All RaMWAS objects except for filematrices are saved to files and loaded back with `saveRDS` and `readRDS` functions.

```{r}
filename = system.file("extdata", "cpgset.rds", package = "ramwas");
cpgset = readRDS(filename);
show(cpgset)

```

# Set of CpG locations

RaMWAS calculates coverage at a precalculated set of CpG. The set depends on the reference genome and usually includes CpGs created by common SNPs. 	The set of CpGs is stored as a `list` with one sorted vector of CpG locations per chromosome.

Here is an examplep of CpG set data structure:

```{r}
cpgset = list( chr1 = c(12L, 57L, 123L),
               chr2 = c(45L, 95L, 99L, 111L),
               chr3 = c(22L, 40L, 199L, 211L) );
```


# Rbam data structure 

RaMWAS scans large BAM files into an economical R representation with `bam.scanBamFile` function.

```{r}
suppressPackageStartupMessages(library(ramwas))
filename = system.file("extdata", "small.bam", package = "ramwas");
rbam = bam.scanBamFile(filename, scoretag = "AS", minscore = 80 ) 
```

The resulting R object is a `list` with 3 components.
First two, `$startsfwd` and `$startsrev`, contain read start positions for forward and reverse strands respectively.

```{r}
head(names(rbam$startsfwd))
rbam$startsfwd$chr2
rbam$startsrev$chr1

```

The `$qc` component holds the quality control measures. 

## Rbam quality control measures

All QC measures are designed to be additive, in the sence that any QC measure calculated for a concatination of two BAM files is equal to the sum of the respective measure calculated for those BAMs separately.

The `small.bam` above has too few reads for meaning QC distributions. Thus we load a small Rbam object with all the QC measures.

```{r}
filename = system.file("extdata", "qc_sample.rds", package = "ramwas");
rbam = readRDS(filename);
```

The currently supported QC measures include th following


### Total number of reads in the BAM file

```{r}
rbam$qc$reads.total
```

### Number of reads in the BAM file aligned to the reference genome

```{r}
rbam$qc$reads.aligned
```

### Number of reads that passed minimum score filter and are recorded in Rbam object

```{r}
rbam$qc$reads.recorded
# sum(sapply(rbam$startsfwd,length)) + sum(sapply(rbam$startsrev,length))
```

### Number of recorded reads aligned to the forward and reverse strands respectively

```{r}
rbam$qc$frwrev
# c(sum(sapply(rbam$startsfwd,length)), sum(sapply(rbam$startsrev,length)))
```

### Distribution of the length of the aligned part of the reads recorded in Rbam

This QC measure is recorded in `$hist.length.matched`. The first element contains the number of reads with 1 aligned basepair, the second with 2, and so on. 
The length of the aligned part of the reads is calculated from the CIGAR strings in the BAM file.

```{r fig.retina=1}
barplot(rbam$qc$hist.length.matched/1e6, width = 1, space = 0, col = "blue", main = "Distribution of the length of the aligned part of the reads", xaxs="i", yaxs="i", ylab = "count, millions");
at = seq(0,length(rbam$qc$hist.length.matched)+10,10);
at[1] = 1;
axis(1,at = at-0.5, labels = at)
```


### Distribution of edit distance between the aligned part of the read and the reference genome

This QC measure is recorded in `$hist.edit.dist1`. The first element contains the number of reads with 0 edit distance (perfect match), the second with edit distance 1, and so on. 

```{r fig.retina=1}
barplot(rbam$qc$hist.edit.dist1/1e6, width = 1, space = 0, col = "blue", main = "Distribution of edit distance", xaxs="i", yaxs="i", ylab = "count, millions");
at = seq(0,length(rbam$qc$hist.edit.dist1)+10,1);
axis(1,at = at+0.5, labels = at)
```

The edit distance is obtained from the NM tag in the BAM file.

### Distribution of the read score (as defined by `scoretag` in the call of `bam.scanBamFile()`)

This QC measure is recorded in `$hist.score1`. The first element contains the number of reads with score of 0, the second with score of 1, and so on. Negative scores are ignored.

```{r fig.retina=1}
barplot(rbam$qc$hist.score1/1e6, width = 1, space = 0, col = "blue", main = "Distribution of read scores", xaxs="i", yaxs="i", ylab = "count, millions");
at = seq(0,length(rbam$qc$hist.score1)+10,20);
axis(1,at = at+0.5, labels = at)
```









